src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                }
+            }
+
+            return ProduceIsSense(Visit(condition), sense);
+        }
+
+        private IOperation ProduceIsSense(IOperation condition, bool sense)
+        {
+            if (!sense)
+            {
+                return new UnaryOperatorExpression(UnaryOperatorKind.Not,
+                                                   condition,
+                                                   isLifted: false, // PROTOTYPE(dataflow): Deal with nullable
+                                                   isChecked: false,
+                                                   operatorMethod: null,
+                                                   semanticModel: null,
+                                                   condition.Syntax,
+                                                   condition.Type,
+                                                   constantValue: default, // revert constant value if we have one.
+                                                   isImplicit: true);
+            }
+
+            return condition;
+        }
+
+        private void VisitConditionalBranch(IOperation condition, ref BasicBlock dest, bool sense)
+        {
+oneMoreTime:
+
+            switch (condition.Kind)
+            {
+                case OperationKind.BinaryOperator:
+                    var binOp = (IBinaryOperation)condition;
+                    bool testBothArgs = sense;
+
+                    switch (binOp.OperatorKind)
+                    {
+                        case BinaryOperatorKind.ConditionalOr:
+                            testBothArgs = !testBothArgs;
+                            // Fall through
+                            goto case BinaryOperatorKind.ConditionalAnd;
+
+                        case BinaryOperatorKind.ConditionalAnd:
+                            if (testBothArgs)
+                            {
+                                // gotoif(a != sense) fallThrough
+                                // gotoif(b == sense) dest
+                                // fallThrough:
+
+                                BasicBlock fallThrough = null;
+
+                                VisitConditionalBranch(binOp.LeftOperand, ref fallThrough, !sense);
+                                VisitConditionalBranch(binOp.RightOperand, ref dest, sense);
+
+                                if (fallThrough != null)
+                                {
+                                    AppendNewBlock(fallThrough);
+                                }
+                            }
+                            else
+                            {
+                                // gotoif(a == sense) labDest
+                                // gotoif(b == sense) labDest
+
+                                VisitConditionalBranch(binOp.LeftOperand, ref dest, sense);
+                                condition = binOp.RightOperand;
+                                goto oneMoreTime;
```

Comment:
Why `goto oneMoreTime;` rather than `VisitConditionalBranch(binOp.RightOperand, ref dest, sense);`? Same question for `UnaryOperator` case below.

Reply:
To avoid unnecessary recursion on deeply nested expressions. Following the pattern used in IL-gen

---
In reply to: [162426752](https://github.com/dotnet/roslyn/pull/24263#discussion_r162426752) [](ancestors = 162426752)